src/main/java/tech/aroma/data/sql/SQLUserPreferencesRepository.kt

```diff
@@ -0,0 +1,175 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tech.aroma.data.sql
+
+import org.slf4j.LoggerFactory
+import org.springframework.dao.EmptyResultDataAccessException
+import org.springframework.jdbc.core.JdbcOperations
+import tech.aroma.data.UserPreferencesRepository
+import tech.aroma.data.assertions.RequestAssertions.validMobileDevice
+import tech.aroma.data.assertions.RequestAssertions.validUserId
+import tech.aroma.data.sql.SQLStatements.*
+import tech.aroma.thrift.channels.MobileDevice
+import tech.aroma.thrift.exceptions.InvalidArgumentException
+import tech.aroma.thrift.exceptions.OperationFailedException
+import tech.sirwellington.alchemy.arguments.Arguments.checkThat
+import tech.sirwellington.alchemy.thrift.ThriftObjects
+import javax.inject.Inject
+
+
+/**
+ *
+ * @author SirWellington
+ */
+internal class SQLUserPreferencesRepository
+@Inject constructor(val database: JdbcOperations, val serializer: DatabaseSerializer<MutableSet<MobileDevice>>) : UserPreferencesRepository
+{
+    override fun saveMobileDevice(userId: String, mobileDevice: MobileDevice)
+    {
+        checkUserId(userId)
+        checkMobileDevice(mobileDevice)
+
+        val sql = Inserts.USER_DEVICE
+        val serialized = ThriftObjects.toJson(mobileDevice)
+
+        try
+        {
+            database.update(sql, userId.toUUID(), serialized)
+        }
+        catch (ex: Exception)
+        {
+            failWithError("Failed to save a new mobile device for [$userId]", ex)
+        }
+    }
+
+    override fun saveMobileDevices(userId: String, mobileDevices: MutableSet<MobileDevice>)
+    {
+        checkUserId(userId)
+        mobileDevices.forEach(this::checkMobileDevice)
+
+        val sql = Inserts.USER_DEVICES
+        val serialized = mobileDevices.map(this::deviceToJson).filterNotNull()
+
+        try
+        {
+            database.update(sql,{ preparedStatement ->
+                preparedStatement.setObject(1, userId.toUUID())
+
+                val serializedStringArray = serialized.toTypedArray()
+                val serializedDevices = preparedStatement.connection.createArrayOf("TEXT", serializedStringArray)
+                preparedStatement.setArray(2, serializedDevices)
+            })
+        }
+        catch (ex: Exception)
+        {
+            failWithError("Failed to save ${mobileDevices.size} devices for [$userId]", ex)
+        }
+    }
+
+    override fun getMobileDevices(userId: String): MutableSet<MobileDevice>
+    {
+        checkUserId(userId)
+
+        val sql = Queries.SELECT_USER_DEVICES
+
+        return try
+        {
+            database.queryForObject(sql, serializer, userId.toUUID()) ?: mutableSetOf()
+        }
+        catch (ex: EmptyResultDataAccessException)
+        {
+            mutableSetOf()
```

Comment:
I didn't see a test case for this.
![bitmoji](https://render.bitstrips.com/v2/cpanel/9284933-288439948_4-s1-v1.png?transparent=1&palette=1&width=246)